src/RegionHintRecommender.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2015 - 2023, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "RegionHintRecommender.hpp"
+#include "RegionHintRecommenderImp.hpp"
+
+#include <cmath>
+#include <fstream>
+
+#include "geopm/Exception.hpp"
+#include "geopm/Helper.hpp"
+
+namespace geopm
+{
+    std::unique_ptr<RegionHintRecommender> RegionHintRecommender::make_unique(const std::string fmap_path, int min_freq, int max_freq)
+    {
+        return geopm::make_unique<RegionHintRecommenderImp>(fmap_path, min_freq, max_freq);
+    }
+
+    RegionHintRecommenderImp::RegionHintRecommenderImp(const std::string fmap_path, int min_freq, int max_freq)
+        : m_min_freq(min_freq)
+          , m_max_freq(max_freq)
+    {
+        std::string buf, err;
+
+        std::ifstream ffile(fmap_path);
+        if (!ffile) {
+            throw Exception("Unable to open frequency map file.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        ffile.seekg(0, std::ios::end);
+        std::streampos length = ffile.tellg();
+        ffile.seekg(0, std::ios::beg);
+
+        std::string fbuf;
+        fbuf.reserve(length);
+
+        std::getline(ffile, fbuf, '\0');
+
+        json11::Json fmap_json = json11::Json::parse(fbuf, err);
+        
+        if (fmap_json.is_null() || !fmap_json.is_object()) {
+            throw Exception("Frequency map file format is incorrect.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        for (const auto &row : fmap_json.object_items()) {
+            if (! row.second.is_array() || row.second.array_items().size() == 0) {
+                throw Exception("Frequency map file format is incorrect.\n",
+                                       GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            m_freq_map[row.first].resize(row.second.array_items().size());
+            for (int idx=0; idx < (int)m_freq_map[row.first].size(); idx++) {
+                m_freq_map[row.first][idx] = row.second[idx].number_value();
+            }
+        }
+    }
+
+    double RegionHintRecommenderImp::recommend_frequency(std::map<std::string, double> nn_output, double phi) const {
+        for (const auto & [region_name, probability] : nn_output ) {
+            if (std::isnan(probability)) {
+                return NAN;
+            }
+        }
+
+        double ZZ = 0;
```

Comment:
```suggestion
        double zz = 0;
```

Modify code below accordingly